### ISSUE TITLE
incorrect font kerning on Windows produces bad output

### ISSUE THREAD

**Issue by glenviewjeff** [wkhtmltopdf/wkhtmltopdf#1527]
Take the following page: http://www.myrecipes.com/recipe/chicken-braised-wine-rosemary-50400000126800/print/

It's especially apparent when zoomed out. It's as if there's no font smoothing or something.

Rendered with wkhtmltopdf:
![Rendered with wkhtmltopdf](https://f.cloud.github.com/assets/334914/2166674/68d2d9ca-950a-11e3-96a6-f2053baec24e.png)

Rendered with Acrobat:
![Rendered with Acrobat](https://f.cloud.github.com/assets/334914/2166678/76ccacd6-950a-11e3-96eb-f8db9f71bc8d.png)

Rendered with wkhtmltopdf:
![Rendered with wkhtmltopdf](https://f.cloud.github.com/assets/334914/2166612/4cf671d6-9509-11e3-8c3c-2828fb479825.png)

Rendered with Acrobat:
![Rendered with Acrobat](https://f.cloud.github.com/assets/334914/2166645/ed04da46-9509-11e3-867a-269bed23852a.png)

**Comment by ashkulz** [comment]
Can you verify if the output better on `0.13-alpha` available from the [downloads page](http://wkhtmltopdf.org/downloads.html)?

**Comment by glenviewjeff** [comment]
Definitely not better, just tested.  Do you want images?

**Comment by ashkulz** [comment]
Yeah, but it looks like a Qt issue because `0.13` is built with Qt 5.4.0 (which is the latest release).

**Comment by glenviewjeff** [comment]
Actually I should retract my statement.  I forgot this was about kerning,
and there doesn't look like a kerning problem here, only a poorly
anti-aliased bold font in wkhtmltopdf in this case.  Interesting thing is
that there's some weird artifacting on the letter el in the word "like"
from the recipe summary, various words in the photo caption, and in the
word "APRIL" under the summary, all in the Acrobat Printer Driver version.
This artifacting does not disappear even when zooming in very close.  So at
least Adobe's got some rendering issues too.  The anti-aliasing seems poor
in the chrome save as pdf file as well.

Browser: http://i.imgur.com/FI1C0MC.png
wkhtmltopdf 0.13-alpha: http://i.imgur.com/qIjfxX7.png
Acrobat Printer Driver: http://i.imgur.com/anapsN9.png
Chrome's built-in "print as pdf": http://i.imgur.com/Avx9J5f.png

**Comment by glenviewjeff** [comment]
A couple of other things, I realized I didn't include the "mashed potatoes" text in the screenshot.  It's kerned correctly in the 0.13-alpha rendering.  I see no kerning issues, but subjectively the quality of the rendering is the worst out of the four.

**Comment by rstarkov** [comment]
The kerning in 0.13-alpha has certainly changed, but I wouldn't say it has improved. Previously the letters were mostly too close, with irregular spacing. Now they are too far, also with irregular spacing.

Comparison: http://i.imgur.com/ud3T5RA.png

I appreciate this is a Qt thing, thanks for trying to fix it but looks like Qt still can't do this right on Windows.

**Comment by mart9634** [comment]
I found a very nice workaround for this issue which might also shed some light on the problem (using wkhtmltopdf 0.12.2.1).

Firstly I saw in some other thread that SVG fonts don't have these kerning issues which is correct. The problem is that the resulting PDF is much larger and the text in it isn't selectable, so using SVG wasn't a viable workaround for me.

The trick is to convert the SVG font back to WOFF, which I did using [Online Font Converter] (http://onlinefontconverter.com/). This produces a WOFF that has perfect kerning just like the SVG. The only strange thing is the line height of the generated WOFF font is not the same as the original and can be many times higher than it should be, but this can be solved using line-height CSS (I just added line-height:1.25em everywhere I set the font-size). It might be possible to fix it by tinkering with the SVG instead but the CSS fix was good enough for me.

One font I used for testing was [Century Gothic from dailywf.com] (http://www.dailywf.com/century_gothic.html). Using their WOFF or TTF directly has kerning issues, with the most notable issue being a huge gap after the letter w. However the WOFF font generated by converting the SVG font using the Online Font Converter has perfect kerning, but with a huge line-height. I also tried converting the Century Gothic TTF from the Windows fonts folder to SVG and then converting the result of that to WOFF and that also works well, with a more reasonable line-height but still slightly higher than normal. Note that converting the Windows TTF directly to WOFF doesn't fix the kerning, it has to go through SVG first.

The WOFF fonts generated by converting SVG fonts aren't good for the screen, so I include a fonts.css file on the wkhtmltopdf command line that contains @font-face declarations for all the affected fonts. These fonts can be named the same as local fonts and wkhtmltopdf will use the ones you supply, so you don't have to mess with your font names. Lastly, wkhtmltopdf doesn't like loading local URL assets from the command-line specified CSS file, so the @font-face declarations have to use data URLs.

Someone with more knowledge of the WOFF format might be able to compare the WOFF files generated by converting TTF and SVG and see why the former has kerning problems and the latter doesn't.

I hope this information helps someone!

**Comment by ashkulz** [comment]
@mart9634: can you post an example HTML/CSS/command-line with your workflow?

**Comment by mart9634** [comment]
@ashkulz: Some sample files are available here:
http://wildwildweb.com.au/temp/wkhtmltopdf-workaround-demo.zip
Not sure how I'm meant to create persistent attachments on this site so it's only on my temporary web server.

Here's the full workflow using the Windows Century Gothic font as an example:

1. Convert C:\Windows\Fonts\GOTHIC.TTF to SVG using [Online Font Converter] (http://onlinefontconverter.com/). This produces the file CenturyGothic.svg. Normally you'd also do the same for italic and bold variants.

2. Convert the CenturyGothic.svg file generated in step 1 to CenturyGothic.woff using the same Online Font Converter.

3. Convert the CenturyGothic.woff file to a data URL using an online tool such as [DATAURL.NET] (http://dataurl.net/#dataurlmaker). Note that most tools don't correctly identify the MIME type so need to replace application/octet-stream with application/font-woff in the output.

4. Create a CSS file called fonts-data.css that contains the @font-face definition of the font(s). Included in zip file.

5. Use the font-family:'Century Gothic' as normal in the CSS referenced from your HTML, but include line-height:1.25em every time you set the font-size. See test.html in the zip file.

To use the demo in the zip file:
1. Extract the zip to a folder
2. Copy wkhtmltopdf.exe (version 0.12.2.1) into the same folder
3. Modify the generate-pdf.bat file to specify the URL path to the folder
4. Run generate-pdf.bat. It should create two output files, one using the workaround and one without.

The letters in the PDF without the workaround are too big and all squished together.
Note that the output files I included in the attached zip is likely to render smaller than the output on other computers as I have Windows set to scale fonts at 150%.
Unfortunately wkhtmltopdf takes the Windows setting into account (making fonts smaller by the same factor) which makes it impossible to get consistent results across different machines, and causing production systems to break if an administrator comes along and changes the Windows font scaling setting.

I also included a file called wkhtmltopdf.css in the zip file. I normally include that file on the command line which includes the fonts-data.css file but I couldn't get it to work in this demo. Yet another wkhtmltopdf bug but not relevant to this issue.

**Comment by johannesspohr** [comment]
@mart9634 Thanks for the detailed post, i tried to follow your instructions for Arial and it really improved the output, but to me the kerning still seems wrong. Looking at "Gothic" in your demo, the "o" is too far on the left and the word "This" seems a lot denser than "Century" (compared it to a browser rendering which looks more balanced in my opinion).

It's a pity that the two most promising free html2pdf approaches (wkhtmltopdf and phantomjs) both can't solve this issue, which makes them effectively unusable for professional applications.

**Comment by mart9634** [comment]
@johannesspohr You're right I hadn't noticed is still isn't quite right. That problem seems to be only with small fonts. The font in my sample PDF is the equivalent of 6pt because my display is set to 150% scaling. There are a couple of ways you can improve kerning of smaller fonts:

* Remove the --disable-smart-shrinking option and use --zoom 1.25. I'm not actually sure what smart shrinking is but it seems to shrink everything by a factor of 1.25. When you correct that by zooming 1.25 the kerning in my example is better.

* If your professional application is back-end, set the windows scaling on the server to something very high like 200%. Then you can set --zoom 2.0 (or 2.5 if you combine it with the first option above) to get even more improvement.

If your professional application is front-end and you're concerned about precision kerning issues then wkhtmltopdf isn't of much use to you, simply because results depend on the Windows font scaling setting so you'll never be able to produce consistent results on all installations.

**Comment by johannesspohr** [comment]
@mart9634 Thanks, it's backend, and after deploying it to the production system (ubuntu 14) i noticed it works perfectly on it (my local installation runs on windows 7 x64). Weird problem though, i hope it will be fixed to work on any platform and zoom level.

**Comment by andrew-pause** [comment]
@mart9634 I find that it looks much better converting directly from ttf to woff, I suspect some kerning data is lost in the svg translation

**Comment by e-dot** [comment]
I had the same issue on Windows with latest wkhtmltopdf (0.12.4), and the only workaround was to **increase the dpi to 200 or more** (I picked 300).

Here's the simple test case:
wkhtmltopdf.letter.spacing.issue.html (compressed)
[wkhtmltopdf.letter.spacing.issue.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/1501967/wkhtmltopdf.letter.spacing.issue.zip)

Calling wkhtmltopdf with no options:
`"C:\Program Files\wkhtmltopdf\bin\wkhtmltopdf.exe" wkhtmltopdf.letter.spacing.issue.html wkhtmltopdf.letter.spacing.issue.pdf`

Here's the result:
[wkhtmltopdf.letter.spacing.issue.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/1501973/wkhtmltopdf.letter.spacing.issue.pdf)

Zooming on a word, you see the issue:
![wkhtmltopdf letter spacing issue zoom on normal](https://user-images.githubusercontent.com/986013/33215684-ded9319e-d130-11e7-8079-2a9fa340728c.png)

Calling the program with option --dpi 300 produces a good workaround:
`"C:\Program Files\wkhtmltopdf\bin\wkhtmltopdf.exe" --dpi 300 wkhtmltopdf.letter.spacing.issue.html wkhtmltopdf.letter.spacing.issue.dpi.300.pdf`

Here's the result:
[wkhtmltopdf.letter.spacing.issue.dpi.300.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/1501975/wkhtmltopdf.letter.spacing.issue.dpi.300.pdf)

Zooming on the same word, it's much better:
![wkhtmltopdf letter spacing issue dpi 300 zoom on normal](https://user-images.githubusercontent.com/986013/33215747-29a1e068-d131-11e7-914d-4aedba80495d.png)

I'm fine using 300 dpi - we generate PDF for printing - but this font rendering issue is a pain. 
Hope this could help,

Regards,

E.

**Comment by PhilterPaper** [comment]
I'm not sure I would call this a "kerning" issue... it looks to me more like poor character width management. Note in the previous post that they appear to be using the same typeface, but something's wrong with where the next glyph starts, even on letters where you would normally expect no kerning adjustments (such as "m" followed by "a"). Could WebKit be using different font width tables than the PDF reader is (same typeface, but different widths)? Is it possible to embed the font in the PDF, and see if that changes anything?

**Comment by pierluca** [comment]
Due to the behaviour of wkhtmltopdf with dynamically created SVGs with foreignObjects containing HTML, I'm working on a project where we can't increase the DPI without messing up the page entirely, as not all elements are scaled in the same way.

Furthermore, the project runs in the cloud and the GDI APIs aren't available, which restricts us from using custom fonts. As such, we're stuck with a kerning issue and a horrible letter-spacing workaround.

Is there any plan to fix this issue in the coming months ? Any pointer as to how to fix it ? We could invest some time, but we don't really know where to start. It's a bit of a deal breaker for us.

**Comment by MartyMcMartface** [comment]
@pierluca I ended up switching to a different library https://github.com/danfickle/openhtmltopdf which doesn't have any of these issues.

**Comment by pierluca** [comment]
@MartyMcMartface That won't cut it for us as we're rendering an AngularJS application which dynamically generates the SVG content :-/

**Comment by pierluca** [comment]
I might have just found out some important complementary information to what @MartyMcMartface mentioned, which is the solution we went with.

We were in the situation where Windows fonts were working more or less correctly but not free fonts. This wasn't acceptable.

I compared the svg generated from Windows TTF fonts and another SVG font (Liberation Sans) - converted to WOFF and then DataURL - that resulted in the wrong kerning.

Here's how I changed Liberation Sans for it to work (some changes may not be needed for it to work, I just went the full way):
* added "glyph-name" attribute to all glyphs
* use hkern g1 and g2 attributes (corresponding to glyph-name) rather than u1 and u2
* glyph with no corresponding unicode were removed from the SVG
* provided full font-face description (all attributes)

It's not perfect but this actually works !

**Comment by xmoles** [comment]
Sorry but i have the same problem 
Why happent this problem 
[farhi_23-10-2018_18-02-35.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2592975/farhi_23-10-2018_18-02-35.pdf)
